dynamax/nonlinear_gaussian_ssm/inference_ekf.py

```diff
@@ -128,12 +132,28 @@ def _step(carry, t):
         # Predict the next state
         pred_mean, pred_cov = _predict(filtered_mean, filtered_cov, f, F, Q, u)
 
-        return (ll, pred_mean, pred_cov), (filtered_mean, filtered_cov)
+        # Build carry states
+        carry = {
+            "marginal_loglik": ll,
+            "filtered_means": filtered_mean,
+            "filtered_covariances": filtered_cov,
+        }
+        carry = {key: val for key, val in carry.items() if key in outputs}
+
+        return (ll, pred_mean, pred_cov), carry
 
     # Run the extended Kalman filter
     carry = (0.0, params.initial_mean, params.initial_covariance)
-    (ll, _, _), (filtered_means, filtered_covs) = lax.scan(_step, carry, jnp.arange(num_timesteps))
-    return PosteriorGSSMFiltered(marginal_loglik=ll, filtered_means=filtered_means, filtered_covariances=filtered_covs)
+    (ll, filtered_mean, filtered_covariance), hist = lax.scan(_step, carry, jnp.arange(num_timesteps))
+
+    # Create posterior object with either final state or states in the history
+    posterior_filtered = {
+        "marginal_loglik": ll,
+        "filtered_means": filtered_mean,
+        "filtered_covariances": filtered_covariance,
+    }
+    posterior_filtered = {**posterior_filtered, **hist}
```

Comment:
The behavior of this is ambiguous, since both `hist` and `posterior_filtered` may have keys with the same name.
Maybe we should write
```
if isfield(history, 'means'):
        filtered_means = hist.filtered_mean
else:
    filtered_mean = filtered_mean

if isfield(history, 'covariances'):
        filtered_covs = hist.filtered_covariances
else:
    filtered_covs = filtered_covariance

if isfield(history, 'loglik'):
        logliks = hist.loglik
else:
    logliks = ll

 return PosteriorGSSMFiltered(
marginal_loglik=logliks,
filtered_means=filtered_means, 
filtered_covariances=filtered_covs)
```